src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtilsUnitTest.java

```diff
@@ -967,4 +968,81 @@ public void testCreateHaplotypeMapping(final VariantContext vc, final Set<Haplot
             Assert.assertEquals(actual.iterator().next(), expected);
         }
     }
+
+    @Test
+    public void testAddGivenHaplotypes() {
+        final int assemblyRegionStart = 1;
+        final int maxMnpDistance = 0;
+        final SmithWatermanAligner aligner = SmithWatermanAligner.getAligner(SmithWatermanAligner.Implementation.FASTEST_AVAILABLE);
+        final AssemblyResultSet assemblyResultSet = new AssemblyResultSet();
+
+        final Haplotype refHaplotype = new Haplotype("AAAACCCCGGGGTTTT".getBytes(), true);
+        final byte[] fullReferenceWithPadding = ("A" + refHaplotype.getBaseString()).getBytes();
+        refHaplotype.setAlignmentStartHapwrtRef(assemblyRegionStart);
+        refHaplotype.setCigar(new Cigar(Collections.singletonList(new CigarElement(refHaplotype.length(), CigarOperator.M))));
+        refHaplotype.setGenomeLocation(new SimpleInterval("chr", assemblyRegionStart, assemblyRegionStart + refHaplotype.length()));
+        assemblyResultSet.setPaddedReferenceLoc(new SimpleInterval("chr", 1, assemblyRegionStart + refHaplotype.length()));
+        assemblyResultSet.add(refHaplotype);
+        assemblyResultSet.setFullReferenceWithPadding(fullReferenceWithPadding);
+
+        // add a SNP
+        final VariantContext givenVC = new VariantContextBuilder("test", "chr", 2, 2,
+                Arrays.asList(Allele.create((byte) 'A', true), Allele.create((byte) 'C', false))).make();
+
+        AssemblyBasedCallerUtils.addGivenHaplotypes(assemblyRegionStart, Collections.singletonList(givenVC), maxMnpDistance,
+                aligner, refHaplotype, assemblyResultSet);
+        Assert.assertEquals(assemblyResultSet.getHaplotypeCount(), 2);
+        Assert.assertEquals(assemblyResultSet.getHaplotypeList().get(1).getBaseString(), "ACAACCCCGGGGTTTT");
+
+
+        // adding the same VC should have no effect
+        AssemblyBasedCallerUtils.addGivenHaplotypes(assemblyRegionStart, Collections.singletonList(givenVC), maxMnpDistance,
+                aligner, refHaplotype, assemblyResultSet);
+        Assert.assertEquals(assemblyResultSet.getHaplotypeCount(), 2);
+
+        // add another SNP
+        final VariantContext givenVC2 = new VariantContextBuilder("test", "chr", 5, 5,
+                Arrays.asList(Allele.create((byte) 'C', true), Allele.create((byte) 'G', false))).make();
+        AssemblyBasedCallerUtils.addGivenHaplotypes(assemblyRegionStart, Collections.singletonList(givenVC2), maxMnpDistance,
+                aligner, refHaplotype, assemblyResultSet);
+        // SNP is not found in existing variation, so it's added to the ref and the first SNP
+        Assert.assertEquals(assemblyResultSet.getHaplotypeCount(), 4);
+        Assert.assertEquals(assemblyResultSet.getHaplotypeList().get(2).getBaseString(), "AAAAGCCCGGGGTTTT");
+        Assert.assertEquals(assemblyResultSet.getHaplotypeList().get(3).getBaseString(), "ACAAGCCCGGGGTTTT");
+
```

Comment:
It looks like your code should handle multi-allelic input alleles, but can you add a test please?

Reply:
done